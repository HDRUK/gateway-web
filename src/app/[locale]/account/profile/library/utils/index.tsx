import { IconButton } from "@mui/material";
import { createColumnHelper } from "@tanstack/react-table";
import { LibraryListItem, SelectedLibrary } from "@/interfaces/Library";
import Link from "@/components/Link";
import StyledCheckbox from "@/components/StyledCheckbox";
import TooltipIcon from "@/components/TooltipIcon";
import { CheckIcon, DeleteForeverIcon } from "@/consts/icons";
import { RouteName } from "@/consts/routeName";

const columnHelper = createColumnHelper<LibraryListItem>();

const getColumns = ({
    handleSelect,
    handleRemove,
    selected,
    translations,
}: {
    handleSelect: (data: SelectedLibrary) => void;
    handleRemove: (id: string | number) => void;
    selected: SelectedLibrary;
    translations: { [id: string]: string };
}) => [
    columnHelper.display({
        id: "actions",
        meta: { isPinned: true },
        cell: ({ row }) => {
            const {
                id,
                datasetId,
                name,
                dataCustodianId,
                dataCustodian,
                darEnabled,
                darTemplatePublished,
                cohortEnabled,
            } = row.original;
            return (
                <div style={{ textAlign: "center" }}>
                    <StyledCheckbox
                        checked={selected[id]?.selected}
                        onChange={(_e, value) =>
                            handleSelect({
                                [id]: {
                                    selected: value,
                                    datasetId,
                                    name,
                                    teamId: dataCustodianId,
                                    teamName: dataCustodian,
                                    darEnabled,
                                    darTemplatePublished,
                                    cohortEnabled,
                                },
                            })
                        }
                        size="large"
                        sx={{ p: 0 }}
                        iconSx={{ mr: 0 }}
                    />
                </div>
            );
        },
        header: () => null,
        size: 43,
    }),

    columnHelper.display({
        id: "name",
        cell: ({ row: { original } }) => {
            const { datasetId, name } = original;
            const linkHref = `/${RouteName.DATASET_ITEM}/${datasetId}`;

            return (
                <Link href={linkHref}>
                    <div style={{ textAlign: "center" }}>{name}</div>
                </Link>
            );
        },
        header: () => <span>{translations.name}</span>,
        size: 150,
    }),

    columnHelper.display({
        id: "darEnabled",
        cell: ({
            row: {
                original: { darEnabled },
            },
        }) => (
            <div style={{ textAlign: "center" }}>
                {darEnabled ? <CheckIcon color="primary" /> : "-"}
            </div>
        ),
        header: () => (
            <TooltipIcon
                buttonSx={{ p: 0 }}
                size="small"
                label={translations.darEnabled}
                content={translations.darEnabled}
            />
        ),
        size: 100,
    }),

    columnHelper.display({
        id: "cohortEnabled",
        cell: ({
            row: {
                original: { cohortEnabled },
            },
        }) => (
            <div style={{ textAlign: "center" }}>
                {cohortEnabled ? <CheckIcon color="primary" /> : "-"}
            </div>
        ),
        header: () => <span>{translations.cohortEnabled}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "dataCustodian",
        cell: ({
            row: {
                original: { dataCustodian },
            },
        }) => <div style={{ textAlign: "center" }}>{dataCustodian}</div>,
        header: () => <span>{translations.dataCustodian}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "entityType",
        cell: ({
            row: {
                original: { entityType },
            },
        }) => <div style={{ textAlign: "center" }}>{entityType}</div>,
        header: () => <span>{translations.entityType}</span>,
        size: 100,
    }),

    columnHelper.display({
        id: "delete",
        cell: ({ row: { original } }) => {
            return (
                <div style={{ textAlign: "center" }}>
                    <IconButton onClick={() => handleRemove(original.id)}>
                        <DeleteForeverIcon color="primary" />
                    </IconButton>
                </div>
            );
        },
        header: () => null,
        size: 43,
    }),
];

export { getColumns };
