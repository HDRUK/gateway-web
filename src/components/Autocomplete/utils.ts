import { ValueType } from "./Autocomplete";

const getChipLabel = (
    options: { value: string | number; label: string }[],
    value: ValueType
) => {
    if (typeof value === "string") {
        return value;
    }

    return options.find(option => option.value === value)?.label;
};

export { getChipLabel };
