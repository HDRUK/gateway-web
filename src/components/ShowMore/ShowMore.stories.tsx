import type { Meta, StoryObj } from "@storybook/react";
import Chip from "../Chip";
import ShowMore from "./ShowMore";

const meta: Meta<typeof ShowMore> = {
    component: ShowMore,
    tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof ShowMore>;

export const Default: Story = {
    args: {
        maxHeight: 24,
        children: Array.from({ length: 20 }).map((_, i) => (
            <Chip label={`chip: ${i}`} />
        )),
    },
};
