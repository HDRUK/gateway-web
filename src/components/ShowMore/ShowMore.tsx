import { useRef, useState, useEffect, ReactNode } from "react";
import { Button } from "@mui/material";
import getNodeDimensions from "get-node-dimensions";
import { useTranslations } from "next-intl";
import { ArrowDropDownIcon } from "@/consts/icons";

interface ShowMoreProps {
    children: ReactNode;
    maxHeight: number;
}

const ShowMore = ({ children, maxHeight }: ShowMoreProps) => {
    const t = useTranslations("components.ShowMore");
    const [needsMoreButton, setNeedsMoreButton] = useState(false);
    const [showAll, setShowAll] = useState(false);
    const myRef = useRef(null);

    useEffect(() => {
        if (!myRef.current) return;
        const { height } = getNodeDimensions(myRef.current);
        setNeedsMoreButton(height > maxHeight);
    }, [maxHeight, children]);

    return (
        <div>
            <div
                style={{
                    overflowY: "hidden",
                    maxHeight: showAll ? "none" : maxHeight,
                }}>
                {children}
            </div>
            {needsMoreButton && (
                <Button
                    style={{ whiteSpace: "nowrap" }}
                    size="small"
                    onClick={() => setShowAll(previous => !previous)}
                    endIcon={
                        <ArrowDropDownIcon
                            fontSize="large"
                            sx={{
                                transform: `rotate(${showAll ? 180 : 0}deg)`,
                            }}
                            color="primary"
                        />
                    }
                    variant="link">
                    {showAll ? t("showLess") : t("showMore")}
                </Button>
            )}
            <div ref={myRef} style={{ display: "none" }}>
                {children}
            </div>
        </div>
    );
};

export default ShowMore;
