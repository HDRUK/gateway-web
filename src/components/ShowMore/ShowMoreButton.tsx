import { Button, ButtonProps, SxProps } from "@mui/material";
import { useTranslations } from "next-intl";
import { ArrowDropDownIcon } from "@/consts/icons";

export interface ShowMoreButtonProps extends Omit<ButtonProps, "onClick"> {
    open?: boolean;
    sxButton?: SxProps;
    onClick: (open: boolean) => void;
}

export default function ShowMoreButton({
    open,
    sxButton,
    onClick,
}: ShowMoreButtonProps) {
    const t = useTranslations("components.ShowMore");

    return (
        <Button
            style={{ whiteSpace: "nowrap" }}
            size="small"
            onClick={() => onClick?.(!open)}
            endIcon={
                <ArrowDropDownIcon
                    fontSize="large"
                    sx={{
                        transform: `rotate(${open ? 180 : 0}deg)`,
                    }}
                    color="primary"
                />
            }
            variant="link"
            sx={sxButton}>
            {open ? t("showLess") : t("showMore")}
        </Button>
    );
}
