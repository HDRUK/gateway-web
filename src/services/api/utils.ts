import { NotificationOptions } from "@/interfaces/Api";
import { AxiosResponse } from "axios";
import { Error } from "@/interfaces/Error";
import notificationService from "../notification";

interface ErrorNotificationProps {
    props: NotificationOptions;
    errorResponse: AxiosResponse<Error>;
    method: "delete" | "post" | "put" | "get";
}

const errorNotification = ({
    errorResponse,
    method,
    props,
}: ErrorNotificationProps) => {
    const { t, i18n, ...notificationProps } = props;
    const { data, status } = errorResponse || {};

    const fallbackTitle = i18n.exists(`api:common.error.status.${status}`)
        ? t(`api:common.error.status.${status}`)
        : "There has been an error";

    const title = data?.title || fallbackTitle;
    const message =
        data?.message ||
        t(`api:common.error.${method}.message`, {
            item: props.itemName || "Item",
        });

    notificationService.apiError(message, {
        title,
        message,
        errors: data?.errors,
        ...notificationProps,
    });
};

interface SuccessNotificationProps {
    props: NotificationOptions;
    method: "delete" | "post" | "put";
}

const successNotification = ({ props, method }: SuccessNotificationProps) => {
    const { t, i18n, ...notificationProps } = props;
    const customMessage = `api:${props.localeKey}.success.${method}.message`;
    const shouldOverideMessage = i18n.exists(customMessage);

    const message = shouldOverideMessage
        ? t(customMessage)
        : t(`api:common.success.${method}.message`, {
              item: props.itemName || "Item",
          });

    notificationService.success(message, { ...notificationProps });
};

export { errorNotification, successNotification };
