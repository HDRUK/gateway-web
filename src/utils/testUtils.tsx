import React, { ReactNode } from "react";
import {
    render,
    renderHook,
    RenderHookOptions,
    RenderOptions,
    RenderResult,
    RenderHookResult,
} from "@testing-library/react";
import { SWRConfig } from "swr";
import { CacheProvider, ThemeProvider } from "@emotion/react";
import ActionBarProvider from "@/providers/ActionBarProvider";
import DialogProvider from "@/providers/DialogProvider";
import theme from "@/config/theme";
import createEmotionCache from "@/config/createEmotionCache";

const clientSideEmotionCache = createEmotionCache();

const Wrapper = ({ children }: { children: ReactNode }) => {
    return (
        <SWRConfig
            value={{
                provider: () => new Map(),
            }}>
            <CacheProvider value={clientSideEmotionCache}>
                <ThemeProvider theme={theme}>
                    <ActionBarProvider>
                        <DialogProvider>{children}</DialogProvider>
                    </ActionBarProvider>
                </ThemeProvider>
            </CacheProvider>
        </SWRConfig>
    );
};

interface OptionProps
    extends RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRender = (
    ui: React.ReactElement<
        unknown,
        string | React.JSXElementConstructor<unknown>
    >,
    options?: OptionProps
): RenderResult => {
    const { wrapperProps, ...rest } = options || {};

    return render(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

interface OptionHookProps
    extends RenderHookOptions<
        unknown,
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement,
        HTMLElement
    > {
    wrapperProps: Record<string, unknown>;
}

const customRenderHook = (
    ui: (initialProps: unknown) => unknown,
    options?: OptionHookProps
): RenderHookResult<unknown, unknown> => {
    const { wrapperProps, ...rest } = options || {};
    return renderHook(ui, {
        wrapper: props => <Wrapper {...props} {...wrapperProps} />,
        ...rest,
    });
};

export * from "@testing-library/react";

export { customRenderHook as renderHook };
export { customRender as render };
