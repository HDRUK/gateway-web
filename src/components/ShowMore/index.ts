import ShowMore from "./ShowMore";
import ShowMoreButton from "./ShowMoreButton";

export default ShowMore;

export { ShowMoreButton };
