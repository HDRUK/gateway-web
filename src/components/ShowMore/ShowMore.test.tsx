import React from "react";
import ShowMore from "@/components/ShowMore";
import { render } from "@/utils/testUtils";

describe("ShowMore", () => {
    it("should render component", async () => {
        const wrapper = render(
            <ShowMore maxHeight={24}>
                Vestibulum ultrices purus sit amet cursus gravida. Proin maximus
                porttitor dui, sed lobortis libero ultrices vitae. In a sem at
                erat venenatis rhoncus. Morbi at diam sed risus commodo tempus
                nec ac ligula. Curabitur arcu velit, volutpat in risus sed,
                suscipit commodo nulla. Aenean luctus feugiat eros at laoreet.
                Fusce rhoncus augue nec tellus ultrices, et tempor sapien
                sollicitudin.Vestibulum ultrices purus sit amet cursus gravida.
                Proin maximus porttitor dui, sed lobortis libero ultrices vitae.
                In a sem at erat venenatis rhoncus. Morbi at diam sed risus
                commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in
                risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at
                laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor
                sapien sollicitudin.Vestibulum ultrices purus sit amet cursus
                gravida. Proin maximus porttitor dui, sed lobortis libero
                ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at
                diam sed risus commodo tempus nec ac ligula. Curabitur arcu
                velit, volutpat in risus sed, suscipit commodo nulla. Aenean
                luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus
                ultrices, et tempor sapien sollicitudin.Vestibulum ultrices
                purus sit amet cursus gravida. Proin maximus porttitor dui, sed
                lobortis libero ultrices vitae. In a sem at erat venenatis
                rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula.
                Curabitur arcu velit, volutpat in risus sed, suscipit commodo
                nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus
                augue nec tellus ultrices, et tempor sapien
                sollicitudin.Vestibulum ultrices purus sit amet cursus gravida.
                Proin maximus porttitor dui, sed lobortis libero ultrices vitae.
                In a sem at erat venenatis rhoncus. Morbi at diam sed risus
                commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in
                risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at
                laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor
                sapien sollicitudin.Vestibulum ultrices purus sit amet cursus
                gravida. Proin maximus porttitor dui, sed lobortis libero
                ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at
                diam sed risus commodo tempus nec ac ligula. Curabitur arcu
                velit, volutpat in risus sed, suscipit commodo nulla. Aenean
                luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus
                ultrices, et tempor sapien sollicitudin.
            </ShowMore>
        );

        expect(wrapper.container).toMatchSnapshot();
    });
});
