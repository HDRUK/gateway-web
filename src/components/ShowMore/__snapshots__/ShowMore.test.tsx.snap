// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowMore should render component 1`] = `
<div>
  <div>
    <div
      style="overflow-y: hidden; max-height: 24px;"
    >
      Vestibulum ultrices purus sit amet cursus gravida. Proin maximus porttitor dui, sed lobortis libero ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor sapien sollicitudin.Vestibulum ultrices purus sit amet cursus gravida. Proin maximus porttitor dui, sed lobortis libero ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor sapien sollicitudin.Vestibulum ultrices purus sit amet cursus gravida. Proin maximus porttitor dui, sed lobortis libero ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor sapien sollicitudin.Vestibulum ultrices purus sit amet cursus gravida. Proin maximus porttitor dui, sed lobortis libero ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor sapien sollicitudin.Vestibulum ultrices purus sit amet cursus gravida. Proin maximus porttitor dui, sed lobortis libero ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor sapien sollicitudin.Vestibulum ultrices purus sit amet cursus gravida. Proin maximus porttitor dui, sed lobortis libero ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor sapien sollicitudin.
    </div>
    <div
      style="display: none;"
    >
      Vestibulum ultrices purus sit amet cursus gravida. Proin maximus porttitor dui, sed lobortis libero ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor sapien sollicitudin.Vestibulum ultrices purus sit amet cursus gravida. Proin maximus porttitor dui, sed lobortis libero ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor sapien sollicitudin.Vestibulum ultrices purus sit amet cursus gravida. Proin maximus porttitor dui, sed lobortis libero ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor sapien sollicitudin.Vestibulum ultrices purus sit amet cursus gravida. Proin maximus porttitor dui, sed lobortis libero ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor sapien sollicitudin.Vestibulum ultrices purus sit amet cursus gravida. Proin maximus porttitor dui, sed lobortis libero ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor sapien sollicitudin.Vestibulum ultrices purus sit amet cursus gravida. Proin maximus porttitor dui, sed lobortis libero ultrices vitae. In a sem at erat venenatis rhoncus. Morbi at diam sed risus commodo tempus nec ac ligula. Curabitur arcu velit, volutpat in risus sed, suscipit commodo nulla. Aenean luctus feugiat eros at laoreet. Fusce rhoncus augue nec tellus ultrices, et tempor sapien sollicitudin.
    </div>
  </div>
</div>
`;
